File: src/main/scala/com/yogpc/qp/packet/advquarry/AdvModeMessage.java
package com.yogpc.qp.packet.advquarry;

import java.util.function.Supplier;

import com.yogpc.qp.machines.advquarry.AdvQuarryWork;
import com.yogpc.qp.machines.advquarry.TileAdvQuarry;
import com.yogpc.qp.machines.base.Area;
import com.yogpc.qp.packet.IMessage;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.network.NetworkEvent;

/**
 * To Client Only
 */
public class AdvModeMessage implements IMessage<AdvModeMessage> {
    ResourceLocation dim;
    BlockPos pos;
    CompoundNBT modeNBT;

    public static AdvModeMessage create(TileAdvQuarry quarry) {
        AdvModeMessage message = new AdvModeMessage();
        message.dim = IMessage.getDimId(quarry.getWorld());
        message.pos = quarry.getPos();
        AdvQuarryWork mode = quarry.action();
        Area digRange = quarry.getArea();
        CompoundNBT nbt = new CompoundNBT();
        CompoundNBT areaNbt = Area.areaToNbt().apply(digRange);
        nbt.put("area", areaNbt);
        CompoundNBT modeNbt = mode.serverWrite(new CompoundNBT());
        nbt.put("mode", modeNbt);
        message.modeNBT = nbt;

        return message;
    }

    @Override
    public AdvModeMessage readFromBuffer(PacketBuffer buffer) {
        pos = buffer.readBlockPos();
        dim = buffer.readResourceLocation();
        modeNBT = buffer.readCompoundTag();
        return this;
    }

    @Override
    public void writeToBuffer(PacketBuffer buffer) {
        buffer.writeBlockPos(pos).writeResourceLocation(dim);
        buffer.writeCompoundTag(modeNBT);
    }

    @Override
    public void onReceive(Supplier<NetworkEvent.Context> ctx) {
        IMessage.findTile(ctx, pos, dim, TileAdvQuarry.class)
            .ifPresent(quarry -> ctx.get().enqueueWork(() -> {
                quarry.setArea(Area.areaLoad(modeNBT.getCompound("area")));
                quarry.action_$eq(AdvQuarryWork.load().apply(quarry).apply(modeNBT.getCompound("mode")));
            }));
    }

}

File: src/main/scala/com/yogpc/qp/packet/advquarry/AdvModeMessage.scala
package com.yogpc.qp.packet.advquarry

import com.yogpc.qp.QuarryPlus
import com.yogpc.qp.packet.IMessage
import com.yogpc.qp.tile.TileAdvQuarry
import net.minecraft.client.Minecraft
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.PacketBuffer
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.{Side, SideOnly}

object AdvModeMessage {
    def create(quarry: TileAdvQuarry) = new AdvModeMessage(quarry)
}

/**
  * To Client Only
  */
class AdvModeMessage(quarry: TileAdvQuarry) extends IMessage {
    private[this] var dim = 0
    private[this] var pos = BlockPos.ORIGIN
    private[this] var modeNBT: NBTTagCompound = _
    if (quarry != null) {
        dim = quarry.getWorld.provider.getDimension
        pos = quarry.getPos
        val nbt = new NBTTagCompound
        modeNBT = quarry.digRange.writeToNBT(quarry.mode.writeToNBT(nbt))
    }

    def this() {
        this(null)
    }

    override def fromBytes(buffer: PacketBuffer) = {
        pos = buffer.readBlockPos
        dim = buffer.readInt
        modeNBT = buffer.readCompoundTag
    }

    override def toBytes(buffer: PacketBuffer) = {
        buffer.writeBlockPos(pos).writeInt(dim)
        buffer.writeCompoundTag(modeNBT)
    }

    @SideOnly(Side.CLIENT)
    override def onRecieve(message: IMessage, ctx: MessageContext) = {
        val world = QuarryPlus.proxy.getPacketWorld(ctx.netHandler)
        if (world.provider.getDimension == dim) world.getTileEntity(pos) match {
            case quarry: TileAdvQuarry =>
                Minecraft.getMinecraft.addScheduledTask(IMessage.toRunnable(() => quarry.recieveModeMassage(modeNBT)))
            case _ =>
        }
        null
    }
}
